## Workflows and Software Management

Workflows are the workhorses of modern bioinformatics -- most analyses require researchers to combine computational steps that involve multiple tools and often multiple programming languages.
When multiple steps are combined together to execute a single analysis, this can generically be referred to as workflow.
While a workflow can be manually executed step-by-step, this is both time-consuming and has the potential to introduce unintended errors. 
Automating a workflow using workflow languages can instead ensure that the entire data analysis is documented and repeatable from start to finish. 

A number of tailored workflow systems have emerged that empower researchers to execute scalable workflows. 
Within these systems, the users specify steps using a system-specific syntax. 
The system then represents these steps as a directed graph in which each node is a step in the workflow, and edges connect sequential steps. 
This back-end organization, paired with additional scaffolding, makes workflows encoded in workflow systems automated, scalable, and transferable across systems.
These traits are highly desirable in bioinformatic workflows where many steps commonly produce thousands of intermediate files.


![**Workflow Systems** Bioinformatic workflow systems have built-in functionality that simplifies running analysis pipelines. **A. Samples** Workflow systems enable you to use the same code to run each step on each sample. Samples can be easily added if the analysis expands. **B. Software Management** Integration with software management tools (e.g. conda, sigularity) can automate software installation for each step. **C. Branching, F. Ordering, G. Parallization** Workflow systems ensure tasks
are executed in the correct order for each sample file, and can automatically execute independent steps in parallel. **D.Standard Steps** Many steps are now considered "standard" (e.g. quality control). Workflow languages keep all information for a step together and can be written to enable you to remix and reuse each individual step across pipelines. **E. Rerun as necessary** Workflow systems keep track of which steps executed properly and on which samples, and allow you to rerun failed steps
(or additional steps) rather than reexecuting the entire workflow. **H. Reporting** Workflow languages enable comprehensive reporting on workflow execution and resource utilization by each tool. **I. Portability** Analyses written in workflow languages (with integrated software management) can be run across computing systems without changes to code.](images/workflow_figure.pdf){#fig:workflow}

### Why are workflows now useful? what has changed? (aka why were people not using them before)
 
The need for workflow management systems has increased with the plummeting cost of sequencing and availability of public data.
While workflows are ubiquitous in bioinformatics and scientific computing in general, workflow systems designed by bioinformaticians for bioinformatic tasks are relatively new. 
Prior to the development of robust workflow systems, common tools for scripting a workflow included make, pydoit, or bash scripting. 
These systems required the user to implement substantial scaffolding around core commands on a per-workflow basis. 
Workflow systems have alleviated this overhead and simplified the process of scripting a workflow.

Advances in workflow systems have led to wide-spread community adoption in part attributable to the open science movement CITE
A critical mass of workflow system-specific code has accumulated such that many routine tasks are already encoded and available for others to use CITE 
At the same time, consensus approaches for routine tasks have emerged, further encouraging reuse of existing code CITE

### How will your life be changed with workflows?

Workflow systems have revolutionized computing in data intensive biology.
Pipelines written in a workflow system are better documented, repeatable, transferable, and scalable. 
Because workflow systems generate directed graphs to execute a workflow, relationships between steps need to be precisely and completely specified for a workflow to execute properly.
This leads to two beneficial side effects. 
First, workflows are more likely to be fully enclosed without undocumented steps that are executed by hand. 
Second, workflows become self-documented; the directed graph produced by workflow systems can be exported and visualized, producing a graphical representation of the relationships between all steps in a pipeline (see {@fig:sgc_workflow}).
When a workflow is specified in this way, each step is executed in the proper order and will be rerun if a failure occurs. 
This frees the user from manually keeping track of execution and monitoring of each step.
Fully-contained workflows (when paired with software management, SEE NEXT SECTION) scripted in a workflow system instill confidence that the code will still execute the same set of commands with little investment by the user in weeks, months, or years from the time of writing.
Similarly, the standard syntax used to specify each step in the workflow lends itself to easy reuse in future project. 
This is in part enabled by cross-system compatibility of most workflow systems, which allows users to develop a workflow e.g. locally, and scale it on a cluster or a cloud computer.

Workflow systems contain powerful built-in features for workflow management that are available to users without the need to write additional code.
The workflow system coordinates execution of the full workflow, including parallelization of non-independent steps. 
When a step fails, it optionally continues with independent steps.
The workflow system keeps track of finished files and removes files if there was a problem in execution.
In addition to coordinating runtime behavior, workflow systems can self-monitor and document resource usage, as well as compile reports that document the results of a workflow. 
The minimal overhead associated with implementing these features greatly empowers the user.

### Getting the benefits without having to learn a scriptable workflow system


While the benefits of encoding a workflow in a workflow system are immense, the learning curve associated with implementing complete workflows in a new syntax can be daunting. 
It is possible to obtain the benefits of workflow systems without learning a workflow software.

Many research groups have used workflow software to build user-friendly pipelines that do not require learning or working with the underlying workflow software.
These tools allow users to take advantage of the benefits of workflow software without needing to invest in curating and writing their own pipeline. 
These tools are specified in an underlying workflow language, but the user interacts with a command-line script and configuration file that coordinate and execute the workflow. 
Often times, workflow parameters are exposed to the user, allowing the user to control certain behaviors such as parallelization or "dry-runs" when executing the command-line script.
Some examples include the ATLAS metagenome assembly and binning pipeline (https://github.com/metagenome-atlas/atlas) \cite{kieser2019atlas}, the Sunbeam metagenome analysis pipeline (https://github.com/sunbeam-labs/sunbeam) \cite{clarke2019sunbeam}, the Elvers \textit{de novo} transcriptome and differential expression pipeline (https://github.com/dib-lab/elvers), the dammit eukaryotic transcriptome pipeline (https://github.com/dib-lab/dammit), and the nf-core RNA-seq pipeline (https://github.com/nf-core/rnaseq/). 

Workflow systems are also available as graphical user interface systems.
Websites like Galaxy, Cavatica, and EMBL-EBI MGnify offer online portals in which users build workflows around publicly-available or user-uploaded data \cite{afgan2018galaxy, volchenboum2017data, mitchell2020mgnify}.

### How to learn to use workflows systems?

There are many scriptable workflow systems that offer similar benefits for data intensive biology. 
Given the plethora of choices and the steep learning curve associated with integrating a workflow system into daily workflow management, it can be difficult to decide which workflow system to adopt. 
While there are many workflow softwares to choose from, each software has it own strengths, meaning each software will meet an individuals computing goals differently (see \textbf{Table \ref{tab:workflows}}).
Our lab has adopted Snakemake given its integration with Python and its flexibility to execute code with different languages (e.g. bash and R) and software management tools (see section XXX) \cite{koster2012snakemake}.
Software like Nextflow and Common Workflow Language scale better to pipelines with hundreds of thousands of steps and support containerization more rigidly, making them ideal for production-level pipelines \cite{di2017nextflow, amstutz2016common}. 
There are also language-specific workflow managers, such as ROpenSci's Drake for R \cite{landau2018drake}. 
Further, workflow systems are not necessarily exclusive entities; Snakemake can export pipelines in Common Workflow Language, allowing the same workflow to be fully compatible with two separate workflow systems. 

\begin{table}
\resizebox{\textwidth}{!}{
\begin{tabular}{|c|c|c|}
\hline
Workflow System & Documentation & Example Workflow \\
\hline
Snakemake & https://snakemake.readthedocs.io/ & https://github.com/snakemake-workflows/chipseq \\
\hline
Nextflow & https://www.nextflow.io/ & https://github.com/nf-core/sarek \\
\hline
common workflow language & https://www.commonwl.org/ & https://github.com/EBI-Metagenomics/pipeline-v5 \\
\hline
workflow definition language & https://openwdl.org/ & https://github.com/gatk-workflows/gatk4-data-processing \\
\hline
\end{tabular}}
\caption{\label{tab:workflows} Popular bioinformatics workflow systems, documentation, and example workflows.}
\end{table}

Independent of computational needs, selecting a workflow system with a strong local or online community can facilitate the adoption process. 
These communities can provide support for new users, and have likely generated many open and accessible workflows that can be modified to analyze new data. 

### Alternatives to workflow systems
Workflow systems are not the only option for constructing and executing a workflow.
Workflow automation can be conducted by scripting the ordered execution of each step in a language such as bash.
While command line scripting is an effective solution to coordinate and execute a workflow, workflows automated in this way do not take advantage of the built-in infrastructure in workflow systems.
In our experience, it is more difficult to identify partially-completed workflow steps that produced truncated files, to rerun specific steps in a workflow, and to add additional files to a workflow when using bash scripting.
These shortcomings are magnified when executing workflows on large-scale sequencing datasets.


### Wrangling Scientific Software 

Most workflows rely on multiple software packages to generate final results.
Bioinformatics research software is heterogeneous, where tools are written by different research groups, in different languages, and with varied target audiences.
Each program has a number of other programs it depends upon to function ("dependencies"), and as software changes over time to meet research needs, the results may change, even when run with identical parameters. 
As a result, it is critical to take an organized approach to installing, managing, and keeping track of software and software versions. 
To meet this need, most workflow managers integrate with software management systems like conda, singularity, and docker \cite{analyticsconda, kurtzer2017singularity, merkel2014docker}.

Software management systems perform some combination of software installation, management, and packaging that alleviate problems that arise from dependencies and that facilitate documentation of software versions.
Conda has a emerged as the leading software installation and management solution (\textbf{Figure \ref{fig:conda_figure}}). 
Conda is a package management and environment management system that allows users to install and organize software.
By enabling installation of software from many languages and disciplines, and managing dependency needs and conflicts for those installations, conda brought about a revolution in package management.
Although portions of conda may eventually be superceded by alternative solutions (e.g. https://github.com/QuantStack/mamba), the model of software installation and management established by conda will likely remain.
Alternatively, container solutions like docker and singularity allow for for the entire computational environment to be captured and distributed, including the operating systems. 
This ensures that an environment is completely reproducible across different computer systems, and is common for production workflows.

While package managers and containers greatly increase reprodubility, there are a number of ways to test out software without needing to worry about installation. 
Some software packages are available as web-based tools and through a series of data upload and parameter specifications, allow the user to interact with a tool that is running on a back-end server. 
This approach is ideal for testing a tool prior to installation to determine whether it produces an appropriate or useful output on your data. 
Integrated development environments like PyCharm and RStudio can also manage software installation for the user for language-specific tools. 

\begin{figure}
\includegraphics[width=0.9\textwidth]{figures/conda_figure.pdf}
\caption{\textbf{The conda package and environment manager simplifies software installation and management.}
\textbf{A. Conda Recipe Repositories} Each program distributed via Conda has a "recipe" describing all software dependencies needed for Conda installation (each of which must also be installable via conda). These are stored and managed in separate "channels", some of which specialize (e.g. "bioconda" specializes in bioinformatic software, "r" specializes in R language packages) \cite{gruning2018bioconda}. \\
\textbf{B. Use Conda Environments to Avoid Installation Conflicts}  Conda does not require root privileges for software installation, thus enabling use by researchers working on shared cluster systems. However, even user-based software installation can encounter dependency conflicts. For example, you might need to use python2 to install and run a program (e.g. older scripts written by members of your lab), while also using snakemake to execute your workflows (requires python$\geq$3.5). By installing each program into an isolated "environment" that contains only the software required to run that program, you can ensure all programs will run without issue. Using small, separate environments for your software and building many simple environments to accommodate different steps in your workflow also reduces the amount of time it takes conda to resolve dependency conflicts between different software tools (e.g. "solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file also specifies which Conda channels to search and download programs from. When specified in a YAML file, conda environments are easily transferable between computers and operating systems. Further, because the version of each package installed in an environment is recorded, workflow reproducibility is enhanced.   
}
\label{fig:conda_figure}
\end{figure}



